File: visitor/oc/bytedesk-oc/ui/views/cell/BDMsgNotificationViewCell.m
//
//  KFDSNotificationViewCell.m
//  feedback
//
//  Created by 萝卜丝 on 2018/2/21.
//  Copyright © 2018年 萝卜丝. All rights reserved.
//

#import "BDMsgNotificationViewCell.h"
#import "BDUtils.h"

@interface BDMsgNotificationViewCell()

@end


@implementation BDMsgNotificationViewCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        self.selectionStyle = UITableViewCellSelectionStyleNone;
//        [self setQmui_shouldShowDebugColor:YES];
        self.backgroundColor = [UIColor systemGroupedBackgroundColor];
    }
    return self;
}

//- (void)awakeFromNib {
//    [super awakeFromNib];
//    // Initialization code
//}
//
//- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
//    [super setSelected:selected animated:animated];
//    // Configure the view for the selected state
//}

- (void)initWithMessageModel:(BDMessageModel *)messageModel {
    //
    _messageModel = messageModel;
    [self addSubviews];
    //
    [self setNeedsLayout];
}


- (void)addSubviews {
    
    if (_timestampLabel) {
        [_timestampLabel removeFromSuperview];
        _timestampLabel = nil;
    }

    if (_contentLabel) {
        [_contentLabel removeFromSuperview];
        _contentLabel = nil;
    }
    
    [self.contentView addSubview:self.timestampLabel];
    [self.contentView addSubview:self.contentLabel];
}


- (UILabel *)timestampLabel {
    //
    if (!_timestampLabel) {
        _timestampLabel = [[UILabel alloc] init];
        _timestampLabel.textColor = [UIColor grayColor];
        _timestampLabel.font = [UIFont systemFontOfSize:11.0f];
//        _timestampLabel.canPerformCopyAction = YES;
        _timestampLabel.textAlignment = NSTextAlignmentCenter;
        [_timestampLabel sizeToFit];
    }
    return _timestampLabel;
}

- (UILabel *)contentLabel {
    //
    if (!_contentLabel) {
        _contentLabel = [[UILabel alloc] init];
        _contentLabel.numberOfLines = 0;
        _contentLabel.lineBreakMode = NSLineBreakByWordWrapping;
        _contentLabel.textColor = [UIColor grayColor];
        _contentLabel.font = [UIFont systemFontOfSize:11.0f];
//        _contentLabel.canPerformCopyAction = YES;
        _contentLabel.textAlignment = NSTextAlignmentCenter;
        [_contentLabel sizeToFit];
    }
    return _contentLabel;
}


- (void)layoutSubviews {
    [super layoutSubviews];
    
    [self layoutTimestampLabel];
    [self layoutContentLabel];
}


- (void)layoutTimestampLabel {
    //
    NSString *timestampString = [BDUtils getOptimizedTimestamp:_messageModel.created_at];
    CGSize timestampSize = [timestampString sizeWithAttributes:@{NSFontAttributeName:[UIFont systemFontOfSize:11.0f]}];
    _timestampLabel.frame = CGRectMake((self.bounds.size.width - timestampSize.width - 10)/2, 0.5f, timestampSize.width + 10.0f, timestampSize.height+1);
    [_timestampLabel setText:timestampString];
}


- (void)layoutContentLabel {
    //
    [_contentLabel setText:_messageModel.content];
    CGSize contentSize = [_contentLabel sizeThatFits:CGSizeMake(300, CGFLOAT_MAX)];
//    CGSize contentSize = [_messageModel.content sizeWithAttributes:@{NSFontAttributeName:[UIFont systemFontOfSize:11.0f]}];
//    NSLog(@"contentSize:%fld, %fld", contentSize.width, contentSize.height);
    
    _contentLabel.frame = CGRectMake((self.bounds.size.width - contentSize.width - 10)/2, 20.5f, contentSize.width, contentSize.height);
}




@end

File: visitor/swift/Sources/bytedesk-swift/Views/cell/BDMsgNotificationViewCell.swift
//
//  File.swift
//  
//
//  Created by 宁金鹏 on 2023/9/10.
//

import UIKit

class BDMsgNotificationViewCell: UITableViewCell {
    
    var timestampLabel: UILabel?
    var contentLabel: UILabel?
    var messageModel: BDMessageModel?
        
    // used when the view is instantiated in code.
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        // debugPrint("BDMsgNotificationViewCell \(#function)")
        //
        self.selectionStyle = UITableViewCell.SelectionStyle.none
        self.backgroundColor = UIColor.systemGroupedBackground
        //
        setupView()
    }

    // for when the view is created via Interface Builder.
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        fatalError("init(coder:) has not been implemented")
    }
    
    func initWithMessageModel(_ msgModel: BDMessageModel?) {
        self.messageModel = msgModel
        timestampLabel?.text = msgModel?.createdAt
        contentLabel?.text = msgModel?.content
        //
        setNeedsLayout()
    }
    
    func setupView() {
        // debugPrint("BDMsgNotificationViewCell \(#function)")
        timestampLabel = UILabel()
        timestampLabel?.font = UIFont.systemFont(ofSize: 11)
        timestampLabel?.textColor = UIColor.systemGray
        timestampLabel?.textAlignment = NSTextAlignment.center
        addSubview(timestampLabel!)
        //
        contentLabel = UILabel()
        contentLabel?.numberOfLines = 0
        contentLabel?.lineBreakMode = NSLineBreakMode.byWordWrapping
        contentLabel?.textColor = UIColor.systemGray
        contentLabel?.font = UIFont.systemFont(ofSize: 11)
        contentLabel?.textAlignment = NSTextAlignment.center
        addSubview(contentLabel!)
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        // debugPrint("BDMsgNotificationViewCell \(#function)")
        let createdAtSize = timestampLabel!.sizeThatFits(CGSizeMake(200, CGFLOAT_MAX)) //self.messageModel?.createdAtSize()
        timestampLabel?.frame = CGRect(x: (self.bounds.size.width - createdAtSize.width)/2, y: 0.5, width: createdAtSize.width, height: createdAtSize.height)
//        let contentSize = self.messageModel!.contentSize()
        let contentSize = contentLabel!.sizeThatFits(CGSizeMake(300, CGFLOAT_MAX))
        //[contentLabel sizeThatFits:CGSizeMake(200, CGFLOAT_MAX)];
        contentLabel?.frame = CGRect(x: (self.bounds.size.width - contentSize.width - 10)/2, y: 20, width: contentSize.width, height: contentSize.height)
    }
    
}
